docs/enhancements/cluster_pool_hot_spares.md

```diff
@@ -0,0 +1,162 @@
+
+2. As a user of a (hypothetical future) ClusterDeployment (or sibling CRD) for a cluster type that doesn't support hibernation,
+
+   I want to be able to configure a ClusterPool that never hibernates its ClusterDeployments
+
+   So that I can use ClusterPools for that kind of ClusterDeployment.
+
+## Proposal
+
+### `spec.runningCount`
+
+Add a field to ClusterPool indicating a portion of the pool to maintain in Running state.
+`spec.runningCount` is an integer absolute number of Running clusters to strive for.
+It is constrained to `0 <= runningCount <= Size`.
+We could enforce this by webhook; or we could just update the spec accordingly.
+
+If unspecified, the default behavior is to hibernate all CDs in the pool, as today.
+This corresponds to `runningCount=0`, conveniently the golang nil value of an int.
+
+### Decouple PowerState on Creation
+
+Today we create all new ClusterDeployments with `powerState=Hibernating` and update that to `Running` at claim time.
+With this feature, we will create with `powerState` unset.
+Then in a separate code path:
+- Sort unassigned (installing + ready) CDs by `creationTimestamp`
+- Update the oldest `runningCount` CDs with `powerState=Running`
+- Update the remainder (the newest `size - runningCount`) with `powerState=Hibernating`
+
+To be safe, and to continue correctly supporting pools with no/zero `runningCount`, we will still always set `powerState=Running` when assigning a claim.
+
+(In both cases, we should only perform an actual `Update()` for a given CD if its state is changed.)
+
+(Note: This means it is possible for a CD's PowerState to be updated several times even while it's installing.
+That's fine -- though we should make sure the hibernation controller handles it sanely.)
+
+### ClusterClaim Assignment by `creationTimestamp`
+
+When determining which ClusterDeployments to assign to new claims, we can simply sort the list by `creationTimestamp`.
+Since we're [prioritizing PowerState the same way](#decouple-powerstate-on-creation), this should serve to satisfy both:
+- FIFO behavior. Oldest clusters will be claimed first.
+- Minimum response time.
+  We always activate the oldest cluster, and we always claim the oldest cluster, so the first claim will pick up the least-recently-resumed cluster, which is the most likely to be Running, or will be there the soonest.
+
+Caveats:
+- Due to variability in install or resume time, it is technically possible that there would be a "better" (faster) choice for a given assignment.
+  But it'll be real close; and we prefer to code more strictly to the FIFO behavior.
+- Manually resuming a pool CD will not cause it to be claimed earlier.
+  In fact, it is subject to being re-hibernated by the controller.
+  This is an acceptable design point: You can manage power states by configuring your ClusterPool; don't expect to be able to do so manually.
+
+### Hibernation Controller Must Account For `hibernateAfter` in Pool Clusters
```

Comment:
what if we look at a general example?

the clusterdeployment starts it's life without hibernateAfter but later on at some point the owner adds the hibernateAfter ..
should the cluster shutdown from the time the hibernateAfter was set or from the last running/installed?

if we think the first hibernation should happen after the duration when hibernateAfter was set, i think we should move towards that change and have the clusterpool set the hibernateAfter when the cluster was assigned to the claim.

Reply:
> I still hold that this entails a mostly-unrelated change to the design/contract of hibernateAfter and is out of scope for this design.

the current recommendation includes updating the hibernation controller to understand what clusters are claimed/or not. I don't think the hibernation controller will be well served to understand a concept that it technically has no reason knowing except that the clusterpool controller sets hibernateAfter but expects that the timer start when clusterpool controller decide.

my recommendation is that we solve this by updating the hibernation controller to support a behavior in general sense where the user wants the timer for hibernation to start when it is set vs always bounded on installation complete of the cluster.

and the clusterpool controller use that behavior to set the hibernateAfter when it is assigning that cluster to a claim.

Reply:
I can get behind the idea of waiting until we assign the CD to set its `hibernateAfter`.

I'm still not seeing a (non-ugly) way of keeping track of when that field was set -- or, for that matter, edited. E.g. if I set the field to 4h, then an hour goes by, then I edit the field from 4h to 2h, do I want my cluster to hibernate in 1h (2h from when the field was initially set) or 2h (2h from when it was updated)? And again, any answer that isn't "2h from installation / the last time the cluster woke up" (the current behavior) represents an API change, and I don't know that we can get away with that.

Suggestions?

Reply:
For now, solving the issue with some tie between clusterpool and hibernation seems like acceptable.
We can look at hibernateAfter sometime later on, it seems like too much of behavior change for this enhancement.